pkg/vmprovider/providers/vsphere2/virtualmachine/backup.go

```diff
@@ -141,24 +121,71 @@ func BackupVirtualMachine(opts BackupVirtualMachineOptions) error {
-// given resources, or an empty string if the data is unchanged in ExtraConfig.
-func getDesiredResourceYAMLForBackup(
+// getDesiredVMResourceYAMLForBackup returns the encoded and gzipped YAML of the
+// given VM, or an empty string if the existing backup is already up-to-date.
+func getDesiredVMResourceYAMLForBackup(
+	vm *vmopv1.VirtualMachine,
+	ecMap map[string]string) (string, error) {
+	curBackup := ecMap[vmopv1.VMResourceYAMLExtraConfigKey]
+	isUpToDate, err := isBackupVMUpToDate(vm, curBackup)
+	if err != nil || isUpToDate {
+		return "", err
+	}
+
+	// Backup the updated VM's YAML with encoding and compression.
+	vmYAML, err := k8syaml.Marshal(vm)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal VM into YAML %+v: %v", vm, err)
+	}
+
+	return util.EncodeGzipBase64(string(vmYAML))
+}
+
+// isBackupVMUpToDate returns true if all of the following fields of the VM are
+// up-to-date with the existing VM backup:
+// - generation (spec changes); annotations; labels.
+func isBackupVMUpToDate(vm *vmopv1.VirtualMachine, backup string) (bool, error) {
+	if backup == "" {
+		return false, nil
+	}
+
+	backupYAML, err := util.TryToDecodeBase64Gzip([]byte(backup))
+	if err != nil {
+		return false, err
+	}
+
+	backupUnstructured := &unstructured.Unstructured{}
+	decUnstructured := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	if _, _, err := decUnstructured.Decode([]byte(backupYAML), nil, backupUnstructured); err != nil {
+		return false, err
+	}
+
+	if vm.GetGeneration() != backupUnstructured.GetGeneration() ||
+		!equality.Semantic.DeepEqual(vm.Labels, backupUnstructured.GetLabels()) ||
+		!equality.Semantic.DeepEqual(vm.Annotations, backupUnstructured.GetAnnotations()) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// getDesiredAdditionalResourceYAMLForBackup returns the encoded and gzipped
+// YAML of the given resources, or an empty string if the existing backup is
+// already up-to-date (checked by comparing the resource versions).
+func getDesiredAdditionalResourceYAMLForBackup(
 	resources []client.Object,
-	ecResourceKey string,
 	ecMap map[string]string) (string, error) {
-	// Check if the given resources are up-to-date with the latest backup.
-	// This is done by comparing the resource versions of each object UID.
-	var isLatestBackup bool
-	if ecResourceData, ok := ecMap[ecResourceKey]; ok {
-		if resourceToVersion := tryGetResourceVersion(ecResourceData); resourceToVersion != nil {
-			isLatestBackup = true
-			for _, curRes := range resources {
-				if curRes.GetResourceVersion() != resourceToVersion[string(curRes.GetUID())] {
-					isLatestBackup = false
-					break
-				}
-			}
+	curBackup := ecMap[vmopv1.AdditionalResourcesYAMLExtraConfigKey]
+	backupVers, err := getBackupResourceVersions(curBackup)
+	if err != nil {
+		return "", err
+	}
+
+	isLatestBackup := true
+	for _, curRes := range resources {
+		if backupVers[string(curRes.GetUID())] != curRes.GetResourceVersion() {
```

Comment:
Should we also include labels and annotations here to re-create the bootstrap resources as close to the backup?  VM operator doesn't consume these, but it's possible that an automation in customers' env relies on the presence of these labels/annotations.

cc @akutz for this thoughts.

Reply:
resourceVersion should be updated for any object changes, so we be able to capture label/annotation changes here by comparing the resourceVersion.

Reply:
Here is an example diff of adding an annotation `foo:bar` to a Secret object:
```
$ diff --suppress-common-lines /tmp/secret-1.yaml /tmp/secret-2.yaml
6a7
>     foo: bar
24a26,34
>   - apiVersion: v1
>     fieldsType: FieldsV1
>     fieldsV1:
>       f:metadata:
>         f:annotations:
>           f:foo: {}
>     manager: kubectl-annotate
>     operation: Update
>     time: "2024-02-07T15:26:24Z"
27c37
<   resourceVersion: "860371"
---
>   resourceVersion: "1596751"
```

Reply:
One more question: do you see _any_ other reason the resource version of the object could be updated (other than changes to the data)?  I am trying to avoid the same scenarios as VM resources where status updates updated the RV.